src/java.base/share/classes/java/lang/Long.java

```diff
@@ -1701,15 +1690,13 @@ public static long divideUnsigned(long dividend, long divisor) {
      * @since 1.8
      */
     public static long remainderUnsigned(long dividend, long divisor) {
-        if (dividend > 0 && divisor > 0) { // signed comparisons
-            return dividend % divisor;
-        } else {
-            if (compareUnsigned(dividend, divisor) < 0) // Avoid explicit check for 0 divisor
-                return dividend;
-            else
-                return toUnsignedBigInteger(dividend).
-                    remainder(toUnsignedBigInteger(divisor)).longValue();
+        // See Hacker's Delight (2nd ed), section 9.3
+        if (divisor >= 0) {
+            final long q = (dividend >>> 1) / divisor << 1;
+            final long r = dividend - q * divisor;
+            return r - (~(r - divisor) >> Long.SIZE - 1 & divisor);
```

Comment:
Parentheses would be particularly helpful here. I'd certainly have to think hard about the relative precedences of `>>`, `-`, and `&`, whereas I wouldn't have to with:
```
return r - ((~(r - divisor) >> (Long.SIZE - 1)) & divisor);
```

I also think it would be worth adding a comment saying that this is deliberately `>>` not `>>>`, even though we have `>>>` in the divide method. Here we're propagating the sign bit so that `thing & divisor` is either 0 or `divisor` according as `thing` is –1 or 0.

Reply:
Hi Éamonn,

I have no problems adding non-strictly needed parentheses. I usually don't because I'm quite familiar with operator precedence and often forget that other people are not.

I'll add comments for parts, like the one you point out, that are not already discussed at length in the bibliographical reference.

Thanks
Raffaello

Reply:
Hi,

I just committed the changes.

Greetings
Raffaello